Doc/library/os.rst

```diff
@@ -249,6 +246,20 @@ process and user.
       Updated to support :pep:`584`'s merge (``|``) and update (``|=``) operators.
 
 
+.. function:: reload_environ()
+
+   Update :data:`os.environ` and :data:`os.environb` with changes to the
+   current process environment made by :func:`os.putenv`, by
+   :func:`os.unsetenv`, or made outside Python in the same process.
+
+   This function is not thread-safe. Calling it while the environment is
+   being modified in an other thread is an undefined behavior. Reading from
+   :data:`os.environ` or :data:`os.environb`, or calling :func:`os.getenv`
+   while reloading, may return an empty result.
```

Comment:
I tried to capture why this function exists in the first place, rather than just describing what the function does.

I think we should also add when we expect the function to be used (e.g. the tkinter example), but I am struggling to come up with a good wording "Generally, this function should only be needed when ...")

A
```suggestion
   The :data:`os.environ` and :data:`os.environb` objects are a cache of 
   environment variables at the time that Python started.
   As such, changes to the current process environment are not reflected
   if made outside Python, or by :func:`os.putenv` or :func:`os.unsetenv`.
   Use :func:`!os.reload_environ` to update :data:`os.environ` and :data:`os.environb`
   with any such changes to the current process environment.

   .. warning::
      This function is not thread-safe. Calling it while the environment is
      being modified in an other thread is an undefined behavior. Reading from
      :data:`os.environ` or :data:`os.environb`, or calling :func:`os.getenv`
      while reloading, may return an empty result.
```

Reply:
The fact `os.putenv` and `os.setenv` don't update the cache mappings feels like a flaw in the way those functions are defined. Fixing that is out of scope for this issue, though.

Aside from replacing "objects" with "mappings" when describing the caches, I like @AA-Turner's suggested wording.

Reply:
If you want to include a specific example of when this can come up, I'd put it in parentheses immediately after mentioning making changes outside of Python:

```
As such, changes to the current process environment are not reflected
   if made outside Python (for example, when Python is running as an embedded component in a larger application), or by :func:`os.putenv` or :func:`os.unsetenv`.
```